### DISCUSSION TITLE
DotnetRestoreAsync() threw exception ---> System.ComponentModel.Win32Exception: The directory name is invalid.

### DISCUSSION THREAD

**Discussion by rogierverkaik** [ASP-NET-MVC-Boilerplate/Framework#166] [Q&A]
### Question
My unit test looks like this:
```
    [TestClass]
    public class ServiceFabricTemplateTest
    {
        public ServiceFabricTemplateTest() => DotnetNew.InstallAsync<ServiceFabricTemplateTest>("ProjectStructureTemplate.sln").Wait();

        [TestMethod]
        public async Task RestoreAndBuild_IsSuccessful()
        {
            using var tempDirectory = TempDirectory.NewTempDirectory();
            var project = await tempDirectory.DotnetNewAsync("rmservicefabric", "TestProject");
            await project.DotnetRestoreAsync();
            await project.DotnetBuildAsync();
        }
    }
```

When I debug the test, the temporary directory is created successfully, the project is created with the correct name. But it fails at the step project.DotnetRestoreAsync();

```
Message: 
    Test method Templatepack.Tests.ServiceFabricTemplateTest.RestoreAndBuild_IsSuccessful threw exception: 
    Boxed.DotnetNewTest.ProcessStartException: Process start failed. Filename:<dotnet> Arguments:<restore> WorkingDirectory:<C:\Users\Rogier\AppData\Local\Temp\0b400f22e23b4c3694cc21ce17d795f8\TestProject>. ---> System.ComponentModel.Win32Exception: The directory name is invalid.
  Stack Trace: 
    Process.StartWithCreateProcess(ProcessStartInfo startInfo)
    Process.Start()
    ProcessExtensions.StartProcessAsync(String filename, String arguments, String workingDirectory, TextWriter outputTextWriter, TextWriter errorTextWriter, Boolean showShellWindow, CancellationToken cancellationToken)
    --- End of inner exception stack trace ---
    ProcessExtensions.StartProcessAsync(String filename, String arguments, String workingDirectory, TextWriter outputTextWriter, TextWriter errorTextWriter, Boolean showShellWindow, CancellationToken cancellationToken)
    ProcessExtensions.StartAsync(String workingDirectory, String fileName, String arguments, Boolean showShellWindow, CancellationToken cancellationToken)
    ProjectExtensions.AssertStartAsync(String workingDirectory, String fileName, String arguments, Boolean showShellWindow, CancellationToken cancellationToken)
    ProjectExtensions.DotnetRestoreAsync(Project project, Nullable`1 timeout, Boolean showShellWindow)
    ServiceFabricTemplateTest.RestoreAndBuild_IsSuccessful() line 17
    ThreadOperations.ExecuteWithAbortSafety(Action action)
```
Am I doing something wrong?

**Comment by RehanSaeed** [comment]
What happens if you manually run `dotnet restore` from the `C:\Users\Rogier\AppData\Local\Temp\0b400f22e23b4c3694cc21ce17d795f8\TestProject` directory? Is that where you have your solution file or csproj file? Is it a multi-project template?

**Comment by rogierverkaik** [comment]
That works fine though.
But the TestProject.sln solution is not in a separate TestProject directory, but in ..\b400f22e23b4c3694cc21ce17d795f8\ itself. So when I run dotnet restore in this folder. The TestProject.sln is restored successfully.

**Comment by RehanSaeed** [comment]
Is this a single project template? Do you have a solution file in your template? If you can provide a link to your template, that would be helpful.

**Comment by rogierverkaik** [comment]
The tree looks like this. So on top I have a Templatepack solution with tests folder and templates folder. In for now I have only 1 template (sftemplate), but in the future that will be more.

Templatepack
---Templatepack.sln
---Templatepack.csproj
---tests
------TemplatePack.Tests
---------Templatepack.Tests.csproj
---templates
------sftemplate
---------.template.config
---------ProjectStructureTemplate.sln
---------src (..)

The code we use is on a private repo in Azure Devops, so I cannot give you access. Maybe there are other ways to share to you?

**Comment by RehanSaeed** [comment]
Can I suggest, cloning the repo and copying the project into your code to debug the issue? Alternatively, you can use [source link debugging](https://github.com/dotnet/sourcelink#experience-in-visual-studio) too which might be simpler.

**Comment by rogierverkaik** [comment]
Good suggestion, I will check it today.

In the meantime I gave you access to my private github repo.

**Comment by rogierverkaik** [comment]
Ok, I found out what happened.
When creating a project called "TestProject" with `.DotnetNewAsync`, it creates a project in `C:\temp\f22ec07657134d1e88dc16e6ab725043`.
But` .DotnetRestoreAsync()` expect the project in `C:\temp\f22ec07657134d1e88dc16e6ab725043\TestProject`. Which is one level deeper.
When I move manually all files into a `\TestProject` (during debugging) the test passes.
Is there a way to solve this? As I far as I know there is no way to set the path of DotnetRestoreAsync and DotnetBuildAsync?

**Comment by rogierverkaik** [comment]
Fixed! I had to add` "preferNameDirectory":true` into template.json file.
